### PR TITLE
fix: use timestamp from relayer

### PR DESCRIPTION
**Pull request by bilbeyt**
Closes: #2022

### CODE REVIEW

File: beamer/agent/chain.py [hunk 1]
```diff
@@ -632,16 +632,17 @@ def maybe_prove(claim: Claim, context: Context) -> None:
 
     def on_future_done(f: Future) -> None:
         try:
-            f.result()
+            result = f.result()
         except Exception as ex:
             context.logger.error("Optimism prove failed", ex=ex, tx_hash=prove_tx)
         else:
+            timestamp = result.stdout.split()[-1].split(": ")[-1]
```
Comment (GabrielBuragev) [inline beamer/agent/chain.py:639]: Does this mean, take the last line of the `stdout` and split it by ":" and take the last value from the split?

If so, i would advise to be more explicit and try to search through all the `stdout` lines to find an exact match of the line "Proof timestamp:*". before trying to split it by ":". 
If the `prove-op-message` command changes in a way that it prints anything else after the `Proof timestamp` line then this will break without us noticing.

Comment (istankovic) [reply beamer/agent/chain.py:639]: > Does this mean, take the last line of the `stdout` and split it by ":" and take the last value from the split?
> 
> If so, i would advise to be more explicit and try to search through all the `stdout` lines to find an exact match of the line "Proof timestamp:*". before trying to split it by ":". If the `prove-op-message` command changes in a way that it prints anything else after the `Proof timestamp` line then this will break without us noticing.

:100:

File: beamer/tests/agent/unit/test_challenge.py [hunk 2]
```diff
@@ -93,8 +96,13 @@ def test_join_false_claim_challenge_only_when_unfilled(filler):
         assert not claim.transaction_pending
 
 
-@patch("beamer.agent.relayer.run_relayer_for_tx")
+@patch("beamer.agent.chain.run_relayer_for_tx")
```
Comment (istankovic) [inline beamer/tests/agent/unit/test_challenge.py:99]: Why this change?

File: beamer/agent/chain.py (hunk 1 above)
Comment (istankovic) [reply beamer/agent/chain.py:639]: I would also prefer if this timestamp parsing was handled by the `relayer` Python module, inside `run_relayer_for_tx`. Because that is where the interface to the relayer binary is.

**Review by istankovic — CHANGES_REQUESTED** [review]
Needs updating, please see comments.

File: beamer/tests/agent/unit/test_challenge.py (hunk 2 above)
Comment (bilbeyt) [reply beamer/tests/agent/unit/test_challenge.py:99]: mock wasnt working before :D my bad :D

File: beamer/agent/relayer.py [hunk 3]
```diff
@@ -30,7 +30,7 @@ def run_relayer_for_tx(
     privkey: HexBytes,
     tx_hash: HexBytes,
     prove_tx: bool = False,
-) -> None:
+) -> str | None:
```
Comment (istankovic) [inline beamer/agent/relayer.py:33]: Suppose we return `None` from this function and there is no exception.
What would happen in `chain.py`?

File: beamer/agent/relayer.py [hunk 4]
```diff
@@ -60,9 +60,14 @@ def run_relayer_for_tx(
...
-    subprocess.run(
+    result = subprocess.run(
         command_args,
         capture_output=True,
         encoding="utf-8",
         check=True,  # check throws an error right away
     )
+
+    if prove_tx:
+        return result.stdout.split()[-1].split(": ")[-1]
```
Comment (istankovic) [inline beamer/agent/relayer.py:71]: This does not address @GabrielBuragev's concern.

I suggest something like the following (totally untested, please adjust accordingly):
```python
match = re.search('Proof timestamp: (\d+)', result.stdout)
if match is not None:
    return match.group(1)
raise RuntimeError("relayer failed to provide proof timestamp")
```

The reason for the exception being that code in `chain.py` can react to it.

File: beamer/agent/relayer.py (hunk 3 above)
Comment (bilbeyt) [reply beamer/agent/relayer.py:33]: this function is also used by l1 resolution so returning None should be fine.

Comment (bilbeyt) [reply beamer/agent/relayer.py:33]: only one prove_tx condition is determining if it is prove or l1 resolution

File: beamer/tests/agent/unit/test_challenge.py [hunk 5]
```diff
@@ -93,8 +95,10 @@ def test_join_false_claim_challenge_only_when_unfilled(filler):
...
 def test_optimism_prove(_mocked_relayer_call):
+    timestamp = int(time.time())
+    _mocked_relayer_call.return_value = str(timestamp)
```
Comment (istankovic) [inline beamer/tests/agent/unit/test_challenge.py:101]: Nitpick: we should remove the leading underscore from the name, we are actually using it now.

**Review by istankovic — APPROVED** [review]
Looks good, thanks!
